cpp/modmesh/buffer/SimpleCollector.hpp

```diff
@@ -0,0 +1,143 @@
+ *   without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <modmesh/buffer/BufferExpander.hpp>
+#include <modmesh/buffer/SimpleArray.hpp>
+
+#include <limits>
+#include <stdexcept>
+
+namespace modmesh
+{
+
+template <typename T>
+class SimpleCollector
+{
+
+private:
+
+    using array_internal_types = detail::SimpleArrayInternalTypes<T>;
+
+public:
+
+    using value_type = typename array_internal_types::value_type;
+    using expander_type = BufferExpander;
+
+    static constexpr size_t ITEMSIZE = sizeof(value_type);
+
+    explicit SimpleCollector(size_t length)
+        : m_expander(BufferExpander::construct(length * ITEMSIZE))
+    {
+    }
+
+    SimpleCollector()
+        : m_expander(BufferExpander::construct())
+    {
+    }
+
+    SimpleCollector(SimpleCollector const & other)
+        : m_expander(other.m_expander->clone())
+    {
+    }
+
+    SimpleCollector & operator=(SimpleCollector const & other)
+    {
+        if (this != &other)
+        {
+            m_expander = other.m_expander->clone();
+        }
+        return *this;
+    }
+
+    SimpleCollector(SimpleCollector && other)
+        : m_expander(std::move(other.m_expander))
+    {
+    }
+
+    SimpleCollector & operator=(SimpleCollector && other)
+    {
+        if (this != &other)
+        {
+            m_expander = std::move(other.m_expander);
+        }
+        return *this;
+    }
+
+    size_t size() const { return expander().size() / ITEMSIZE; }
+    size_t capacity() const { return expander().capacity() / ITEMSIZE; }
+
+    value_type const & operator[](size_t it) const noexcept { return data(it); }
+    value_type & operator[](size_t it) noexcept { return data(it); }
+
+    void reserve(size_t cap) { expander().reserve(cap * ITEMSIZE); }
+    void expand(size_t length) { expander().expand(length * ITEMSIZE); }
+
+    value_type const & at(size_t it) const
+    {
+        validate_range(it);
+        return data(it);
+    }
+    value_type & at(size_t it)
+    {
+        validate_range(it);
+        return data(it);
+    }
+
+    SimpleArray<T> as_array()
+    {
+        return SimpleArray<T>(expander().as_concrete());
```

Comment:
perhaps have assert have to ensure the underlying buffer size is correct?

Reply:
I don't think you can check for the size correctness here.  The size information is stored and calculated from the `expander()` inside.

Reply:
I am thinking about this:
```
size_t size buffer_size = expander().as_concrete().nbytes();
assert(buffer_size % sizeof(T) == 0)
```

Reply:
It is redundant code, because it is guaranteed to be divisible in `reserve()` and `expand()` https://github.com/solvcon/modmesh/blob/3d1944e2342118b62e3349f425b3585a89002b6c/cpp/modmesh/buffer/SimpleCollector.hpp#L99

Adding the calculation increases unnecessary instructions and misleads maintainers in a way that they might think we don't need performance here.